### PR TITLE
Credit Card forms: Add checkbox that assigns new cards to all subscriptions

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

This adds a "Use this payment method for all subscriptions on my account." checkbox to the Stripe credit card forms in checkout and payment method management. A checkbox for this (with slightly different wording) already exists in the non-subscription "Add Payment Method" form, but this removes it in favor of the new one.

Screen Shot 2021-10-01 at 7 15 24 PM

On the non-subscription "Add Payment Method" form, the checkbox will be pre-checked (as it is currently), but all other instances of the checkbox (checkout and subscription-level Add/Change Payment Method forms) will default to unchecked.

If the new element is checked when each of the relevant forms is submitted, it will replace the payment method on _all_ of the user's active subscriptions with the newly added credit card.

For simplicity during implementation, this functionality is not yet being added to existing credit cards, Ebanx credit cards, PayPal, or any other payment method, but it may be added (particularly to PayPal) in the future.

Requires D67859-code to update the endpoints to support the new property. 

#### Motivation

The reason for this change is that many users come to these forms intending to update an expiring or otherwise incorrect payment method, but may be unaware that they need to make the change to each of their subscriptions. As a result they may update an expiring card on one subscription and then be confused when they receive a notification that another subscription is still using the expiring card.

This is quite common since users often purchase several products at once (e.g.: a plan and a domain) and are not clear that the payment methods for these purchases must be managed individually. While a more advanced payment management interface would be nice, it is our hope that while that project is being discussed, this checkbox will allow such users to update their card once and therefore avoid some confusion.

Part of 509-gh-Automattic/payments-shilling

#### To do

- [x] Add checkbox to all Stripe card forms.
- [x] Remove existing checkbox on add payment method form.
- [x] Send data to payment method add endpoint in place of old checkbox data.
- [x] Send data to transactions endpoint.
- [x] Add real "Learn more" link target to checkbox label.
- [x] Make checkbox pre-checked on non-subscription add payment method form.
- [x] Hide checkbox for non-wpcom purchases.
- [x] Modify transaction endpoint to use the new property.
- [x] Remove CSS for removed checkbox.

#### Screenshots

Here's the old checkbox in the non-subscription "Add Payment Method" form that is being replaced by this PR:

Screen Shot 2021-10-01 at 7 26 58 PM

Here's how the new checkbox looks on the "Add Payment Method" form:

Screen Shot 2021-10-01 at 7 28 54 PM

Here's how the new checkbox looks in checkout:

Screen Shot 2021-10-01 at 4 17 13 PM

#### Testing instructions

To test checkout:

- Use a site that has an existing wpcom purchase (eg: a plan or a domain).
- Add a new product to the cart (if your existing purchase is a plan, this new product cannot also be a plan because it will replace the old one) and visit checkout.
- Fill out the checkout form and select "Credit or Debit card" as your payment method.
- Check the "Use this payment method for all subscriptions on my account." checkbox.
- Click to submit the purchase.
- Visit `/me/purchases` and verify that your existing purchase has been changed to use the new card as its payment method.

To test subscription payment method management:

- First, use a site that has an existing non-wpcom purchase (eg: a Jetpack site).
- Visit `/me/purchases`, select your existing purchase, click "Change payment method", and verify that you _do not_ see the "Use this payment method for all subscriptions on my account." checkbox in the new card form.
- Next, use a site has _more than one_ existing wpcom purchase (eg: a plan or a domain).
- Visit `/me/purchases`, select your existing purchase, and click "Change payment method". 
- Select "Credit or Debit card", fill in the new card form and check the "Use this payment method for all subscriptions on my account." checkbox.
- Click to submit the form.
- Visit `/me/purchases` and verify that all your purchases have been changed to use the new card as their payment method.

To test non-subscription payment method management:

- Use a site has _more than one_ existing wpcom purchase (eg: a plan or a domain).
- Visit `/me/purchases/add-payment-method`.
- Fill in the new card form and verify that the "Use this payment method for all subscriptions on my account." checkbox is pre-checked.
- Click to submit the form.
- Visit `/me/purchases` and verify that all your purchases have been changed to use the new card as their payment method.

### CODE REVIEW

**Comment by nbloomf** [comment]
I haven't fully tested this yet, and we talked about this a bit in slack, but while branching from this PR to add some analytics I noticed I was able to consistently trigger an error using the following steps:

1. Load a subscription level Change Payment Method screen
2. Reload the page

The error raised is as follows:

```
index.jsx:84 Uncaught TypeError: Cannot read properties of undefined (reading 'productSlug')
    at ChangePaymentMethod (index.jsx:84)
    at renderWithHooks (react-dom.development.js:14985)
    at mountIndeterminateComponent (react-dom.development.js:17811)
    at beginWork (react-dom.development.js:19049)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at beginWork$1 (react-dom.development.js:23959)
    at performUnitOfWork (react-dom.development.js:22771)
    at workLoopSync (react-dom.development.js:22702)
```

Edit: looks like this is fixed now!

**Comment by sirbrillig** [comment]
Thanks for noticing that, @nbloomf! After fixing that error I decided to convert `ChangePaymentMethod` to TS to see if there were any other possible issues. There was one possible problem where there's a redirect that wasn't in a `useEffect` and there were also several unused props being passed in, which I was able to remove.

File: client/me/purchases/manage-purchase/payment-method-selector/index.tsx [hunk 1]
```diff
@@ -142,48 +134,12 @@ export default function PaymentMethodSelector( {
...
-		return (
-			<span className="payment-method-selector__all-subscriptions-effect-warning">
-				{ translate( 'This card will be used for future renewals of existing purchases.' ) }
-			</span>
-		);
-	}
-	return (
-		<span className="payment-method-selector__all-subscriptions-effect-warning">
-			{ translate(
-				'This card will not be assigned to any subscriptions. You can assign it to a subscription from the subscription page.'
```
Comment (sirbrillig) [inline client/me/purchases/manage-purchase/payment-method-selector/index.tsx:181]: I wonder if we want to keep this warning around somewhere somehow. It seems valuable for the new-card form.

**Comment by pottedmeat** [comment]
Link target will be https://wordpress.com/support/payment/#using-a-payment-method-for-all-subscriptions which will be moved over from draft once this is depoyed.

**Comment by sirbrillig** [comment]
Testing revealed that the transactions endpoint changes do not work for 3DS cards, so we've disabled that part of this PR for now.

**Comment by nbloomf** [comment]
Looks good! I tested this along with the backend change. My detailed notes are there, but TL;DR I don't see any problems.

**Review by nbloomf — APPROVED** [review]

**Comment by jjchrisdiehl** [comment]
Ran through all three tests.

Test 1 works as expected, I successfully changed the card being used by selecting the checkbox during checkout.

Test 2 didn't work as expected. In the first part I added a Jetpack Security Daily subscription to my personal .org site. When I went to /me/purchases, clicked on the purchase, then clicked on Change Payment Method I saw the checkbox in the 'Credit or debit card' form. In this test the checkbox should not appear.

https://d.pr/i/y34bay

However, part two of Test 2 worked as expected.

Test 3 works as expected as well.

**Comment by sirbrillig** [comment]
> Testing revealed that the transactions endpoint changes do not work for 3DS cards, so we've disabled that part of this PR for now.

After deploying the backend changes, it turns out that 3DS cards in checkout do actually work; the reason it appeared that they didn't was because when checkout completed via webhook, the backend patch wasn't applied. I've re-enabled that feature here.

> Test 2 didn't work as expected. In the first part I added a Jetpack Security Daily subscription to my personal .org site. When I went to /me/purchases, clicked on the purchase, then clicked on Change Payment Method I saw the checkbox in the 'Credit or debit card' form. In this test the checkbox should not appear.

Thanks for noticing that, @jjchrisdiehl ! There was one thing I forgot which was to actually _hide_ the checkbox when it was set to not operate. There was another bug too which made it so that even if the checkbox was supposed to be disabled, it could still be turned on and would submit its data. Both bugs should be fixed now.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6790125

Thank you @sirbrillig for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6790125).

**Pull request merged**